experimental/displaynames/src/displaynames.rs

```diff
@@ -399,90 +418,207 @@ impl LocaleDisplayNamesFormatter {
     );
 
     /// Returns the display name of a locale.
+    /// This implementation is based on the algorithm described in
+    /// https://www.unicode.org/reports/tr35/tr35-general.html#locale_display_name_algorithm
+    ///
     // TODO: Make this return a writeable instead of using alloc
     pub fn of<'a, 'b: 'a, 'c: 'a>(&'b self, locale: &'c Locale) -> Cow<'a, str> {
-        // https://www.unicode.org/reports/tr35/tr35-general.html#Display_Name_Elements
+        let longest_matching_subtag = find_longest_matching_subtag(&locale, &self);
 
-        // TODO: This binary search needs to return the longest matching found prefix
-        // instead of just perfect matches
-        if let Some(displayname) = match self.options.style {
-            Some(Style::Short) => self
-                .locale_data
-                .get()
-                .short_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            Some(Style::Long) => self
-                .locale_data
-                .get()
-                .long_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            Some(Style::Menu) => self
-                .locale_data
-                .get()
-                .menu_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            _ => None,
+        // Step - 1: Construct a locale display name string (LDN).
+        // Find the displayname for the longest_matching_subtag which was derived above.
+        let ldn = get_locale_display_name(&locale, &longest_matching_subtag, &self);
```

Comment:
It might be worth combining these two calls so you don't have to do the map lookup twice. You can add a `&'a str` field to the `LongestMatchingSubtag` enum.

Reply:
Map lookup is O(nlogn), and the `names` map is probably the biggest, so ideally we wouldn't do that lookup twice.